### PR TITLE
ingest: Introduce the dissect processor

### PR DESCRIPTION
**Pull request by jakelandis**
The ingest node dissect processor is an alternative to Grok
to split a string based on a pattern. Dissect differs from
Grok such that regular expressions are not used to split the
string.

Dissect can be used to parse a source text field with a
simpler pattern, and is often faster the Grok for basic string
parsing. This processor uses the dissect library which
does most of the work.

--------

EDIT: removed comments about breaking change, since this will no longer be a breaking change.

Rally Benchmark:
-----------------
Rally track here: https://github.com/jakelandis/rally-tracks-ingest/blob/dissect_tests/http_logs_for_ingest/track.json
Full test results here: https://gist.github.com/jakelandis/de554f394e5455ee66797c2b956d63bf

Single node, 1 shard, 0 replicas, i7 4 core/8 thread running ~3.7GHz with NVMe disk. 

_source is disabled and refresh_interval is set to -1.

All fields use the following mapping to focus on the ingest node performance. 
```
"type": "keyword",
"index": false,
"doc_values": false
```

Data set is the same as data set as https://github.com/elastic/rally-tracks/tree/master/http_logs  based on Web server logs from the 1998 Football world cup. However, this test uses an un-parsed format for the data.

Example log lines:

```
{"message" : "30.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /images/info.gif HTTP/1.0\" 200 1251"}
{"message" : "28.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /french/images/hm_official.gif HTTP/1.1\" 200 972"}
{"message" : "17.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /french/hosts/cfo/images/cfo/cfophot3.jpg HTTP/1.0\" 200 6695"}
{"message" : "72.236.7.0 - - [1998-05-24T15:00:01-05:00] \"GET /images/s102380.gif HTTP/1.1\" 200 207"}
{"message" : "82.36.0.0 - - [1998-05-24T15:00:02-05:00] \"GET /images/nav_bg_bottom.jpg HTTP/1.0\" 200 8389"}
```

The dataset has 247,249,096 documents ... each test was run 4 times, once with 1 lap (~250m), once with 2 laps (~500m), once with 3 laps (~750m), and once with 4 laps (~1b). 

Baseline - runs an ingest pipeline that does nothing (specifically it attempts to upper case a missing field).  This accounts for the overhead of creating the IngestDocument and pushing the data through a pipeline.

Grok - extracts the timestamp, ip, request, status, and size from the source string with the Grok processor. (`%{IPORHOST:clientip} %{HTTPDUSER} %{USER} \\[%{TIMESTAMP_ISO8601:@timestamp}\\] \"(?:%{WORD} %{NOTSPACE:request}(?: HTTP/%{NUMBER})?|%{DATA})\" %{NUMBER:status} (?:%{NUMBER:size}|-)`)

Dissect - extracts the timestamp, ip, request, status, and size from the source string with the Dissect processor. (`%{clientip} %{?ident} %{?auth} [%{@timestamp}] \"%{?verb} %{request} HTTP/%{?httpversion}\" %{status} %{size}`)

Results _(average the medians across all runs)_ 

Dissect is about 8% faster then Grok. (ranges from 7-10% across runs)
Dissect is about 15% slower then the Baseline. (ranges from 14-16% across runs)
Grok is about 22% slower then the Baseline. (ranges from 20%-24% across runs)

Throughput (docs/second - higher is better)
image

----------

Screen shot of rendered help here: https://gist.github.com/jakelandis/dd0b627029f26fd52957a00b0c7e118f

### CODE REVIEW

**Comment by jakelandis** [comment]
@rjernst - the intent here is to match Beats and Logstash's implementation.  However, in the course of implementing for this for the ingest I found a point of confusion w/r/t an overloaded use of the `?` symbol (as well as a couple other minor things) that I wanted to change. Unfortunately this is a breaking w/r/t to the way Beats and Logstash currently work.  

After some discussion with Beats and Logstash, they plan to make a corresponding (breaking) change for 7.x too. However there was a desire to allow copy/pasting of patterns between products withing major versions, which necessitates a breaking change here too.    

> If we must break this, then the first PR should have the intended 6.x behavior, and a followup with the breaking change in 7.0 should be made separately.

Apologies for the mis-ordering of PR's. I had planned once all changes were in master to backport and then add an extra commit the 6.x rules (cherry-picking + 1 commit), and seek review on that one commit.

If we are OK with ingest node dissect syntax being slightly different then Beats and Logstash in 6.x, then I can remove the breaking changes label and do a normal backport. The difference is pretty minor ... the change between 6.x and 7.x is `?` should now be `*` and some extra validation is enforced. The 7x->6x code diff is here: https://github.com/jakelandis/elasticsearch/compare/c2ada878a32ba1d4c7f9b36816fd6492824b484f...2b4a4d58d7bc1fbc9b459a0e06eb1f75c4f1323d (some extra fluff in that diff which will get cleaned up for the PR)

**Comment by rjernst** [comment]
The logic makes sense for why a breaking change. But the order should be as I described in the previous comment. Can this PR be amended with the old behavior?

**Comment by jakelandis** [comment]
@rjernst I am re-visiting the discussion around the need to make this a breaking change. I will update this PR with the result of that discussion.

**Comment by jakelandis** [comment]
After further discussion, this will NOT be a breaking change. The syntax between ingest/Logstash/Beats will have a minor difference w/r/t `*` vs `?` for 6.x...but that is OK since it will avoid a breaking change in Elasticsearch. 

I am removing the breaking label.

@rjernst This is ready to review, and will be a normal cherry-picked backport.

**Comment by jakelandis** [comment]
ping @rjernst

**Review by rjernst — APPROVED** [review]
LGTM, although I am a bit apprehensive about this being in ingest-common? That is turning into what I was afraid of: a catch all where all ingest processors seem to go.

File: docs/reference/ingest/ingest-node.asciidoc [hunk 1]
```diff
@@ -1049,6 +1049,199 @@ understands this to mean `2016-04-01` as is explained in the <<date-math-index-n
...
+    "field": "message",
+    "pattern" : "%{clientip} %{ident} %{auth} [%{@timestamp}] \"%{verb} %{request} HTTP/%{httpversion}\" %{status} %{size}"
+   }
+}
+--------------------------------------------------
+// NOTCONSOLE
+[[dissect-key-modifiers]]
+==== Dissect key modifiers
+Key modifiers can change the default behavior for dissection. Key modifiers may be found on the left or right
+of the `%{keyname}` always inside the `{%` and `}`. For example `%{+keyname ->}` has the append and right padding
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply docs/reference/ingest/ingest-node.asciidoc]: fixed. thanks!

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DissectProcessor.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        super(tag);
+        this.field = field;
+        this.ignoreMissing = ignoreMissing;
+        this.pattern = pattern;
+        this.appendSeparator = appendSeparator;
+        this.dissectParser = new DissectParser(pattern, appendSeparator);
+    }
+
+    //package private for testing
+    String getField() {
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DissectProcessor.java]: yup. removed the getters and made the members package private. thanks!

**Comment by rjernst** [comment]
It might be worth discussing with the core team. Grok was originally a plugin because ingest-common was supposed to have no dependencies.

**Comment by jakelandis** [comment]
All green merging. will backport as-is for inclusion with 6.5

**Pull request merged**